src/core/truthTable/truth_table.cpp

```diff
@@ -0,0 +1,252 @@
+
+namespace syrec {
+
+    std::vector<truthTable::cube_type> in_cube_to_full_cubes(const truthTable::cube_type& c, std::vector<truthTable::cube_type>& result) {
+        auto first = c.begin();
+
+        auto last = c.end();
+
+        truthTable::cube_type cube(first, last);
+
+        std::vector<unsigned> dcPositions;
+
+        unsigned pos = 0;
+
+        while (first != last) {
+            if (!((*first).has_value())) // if DC
+            {
+                std::cout << "dontcare" << std::endl;
+                dcPositions.emplace_back(pos);
+            }
+
+            ++pos;
+
+            ++first;
+        }
+
+        for (unsigned i = 0; i < (1u << dcPositions.size()); ++i) {
+            for (unsigned j = 0; j < dcPositions.size(); ++j) {
+                unsigned localBit = i & (1u << (dcPositions.size() - j - 1)) ? 1 : 0;
+
+                cube.at(dcPositions.at(j)) = localBit;
+            }
+
+            result.emplace_back(cube);
+        }
+
+        return result;
+    }
+
+    truthTable::cube_type number_to_cube(unsigned number, unsigned bw) {
+        truthTable::cube_type c;
+
+        for (unsigned i = 0; i < bw; ++i) {
+            c.push_back((number & (1u << (bw - 1 - i))) ? true : false);
+        }
+
+        return c;
+    }
+
+    void extend_truth_table(truthTable& tt) {
+        std::map<std::vector<truthTable::cube_type>, truthTable::cube_type> newCubes;
+
+        for (auto& it: tt.io_cube()) {
+            std::vector<truthTable::cube_type> inCube;
+
+            in_cube_to_full_cubes(it.first, inCube);
+
+            truthTable::cube_type output(it.second.begin(), it.second.end());
+
+            newCubes.insert(std::make_pair(inCube, output));
+        }
+
+        tt.clear();
+
+        for (auto& it: newCubes) {
+            for (auto& itCube: it.first) {
+                truthTable::cube_type newInEntry = itCube;
+
+                tt.add_entry(newInEntry, it.second);
+            }
+        }
+
+        truthTable::cube_type outCube(tt.num_outputs(), false);
+
+        unsigned currentPos = 0;
+
+        truthTable::cube_vector ioVec = tt.io_cube();
+
+        for (auto it = ioVec.begin();; ++it) {
+            unsigned pos = 0;
+
+            unsigned i = tt.num_inputs();
+
+            if (it == ioVec.end()) {
+                pos = 1u << tt.num_inputs();
+            } else {
+                for (auto& inBit: it->first) {
+                    pos |= (*inBit) << --i;
+                }
+            }
+
+            for (i = currentPos; i < pos; ++i) {
+                truthTable::cube_type inputCube = number_to_cube(i, tt.num_inputs());
+
+                tt.add_entry(inputCube, outCube);
+            }
+
+            currentPos = pos;
+
+            if (it == ioVec.end()) {
+                break;
+            }
+        }
+    }
+
+    truthTable::cube_type append_zero(truthTable::cube_type enc) {
+        enc.push_back(false);
+
+        return enc;
+    }
```

Comment:
this creates a copy of a potentially large vector once. then adds a single element and then makes another copy when returning.

This should be a member function of the `Cube` class that simply appends a zero and does not return anything.